### PR TITLE
Fix issue 6878 | [Bug] ShellItem.Items.Clear() crashes when the ShellItem has bottom tabs

### LINKED ISSUE #6878: [Bug] ShellItem.Items.Clear() crashes when the ShellItem has bottom tabs
### Description
When using shell, if you want to programmatically clear and redraw the bottom tabs on a shell page, you can't use Items.Clear to get rid of all the existing ones because it crashes. Investigation suggests that this is because it's trying to set properties on a tab that is no longer there.

### Steps to Reproduce

1. Create a ShellItem in xaml that contains tabs.
2. In code, call Items.Clear() on your ShellItem.
3. A null exception will occur on line 83 of ShellBottomNavViewAppearanceTracker.SetBackgroundColor, presumably because menuView.GetChildAt(index) returns null, either because index is -1 or menuView has no children.

### Expected Behavior
The tabs are cleared and you're free to create new ones
### Actual Behavior
Exception and crash
### Basic Information

- Version with issue: 4.0.0.425677
- Last known good version: Unknown
- IDE: Visual Studio 2019
- Platform Target Frameworks: 
  - iOS:  Untested
  - Android: 9.0

### PR DESCRIPTION
**Pull request by felipebaltazar**
### Description of Change ###

I followed another PR (#8079) feedback and create a null check to avoid NullException for `ShellBottomNavViewAppearanceTracker.SetBackgroundColor`

### Issues Resolved ### 

- fixes #6878 

### API Changes ###
Changed:
 - void ShellBottomNavViewAppearanceTracker.SetBackgroundColor
	- added this validation:

```C#
	var child = menuView.GetChildAt(index);

	if (child == null)
		return;
```
 
### Platforms Affected ### 
- Core/XAML (all platforms)

### Behavioral/Visual Changes ###

None

### Before/After Screenshots ### 

Not applicable

### Testing Procedure ###
1. Create a ShellItem in that contains tabs and set background-color
2. In code, call Items.Clear() on your ShellItem.
3. No exceptions at this point

- Or just run 6878 issue at Control Galery:

<p align="center">
	<kbd>
		image
	</kbd>
</p>

### PR Checklist ###

- [ ] Targets the correct branch
- [ ] Tests are passing (or failures are unrelated)

### CODE REVIEW

**Comment by felipebaltazar** [comment]
@PureWeen  @rmarinho 
Hello, any position about this PR?

File: Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.Shared/TestPages/TestPages.cs [hunk 1]
```diff
@@ -612,7 +612,7 @@ public ContentPage AddTopTab(string title, string icon = null)
 
 		public void AddTopTab(ContentPage page, string title = null, string icon = null)
 		{
-			if (Items.Count == 0)
+			if(Items.Count == 0 || Items[0].Items.Count == 0)
```
Comment (PureWeen) [inline Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.Shared/TestPages/TestPages.cs]: Is this change needed?

If we're wanting to act on this being true *Items[0].Items.Count == 0* then we would add a ShellSection to Items[0] not call AddContentPage

But when I run your test code this condition is never true so I'm assuming we can just revert this change?

Comment (felipebaltazar) [reply Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.Shared/TestPages/TestPages.cs]: I've changed this to use Items[0].Clear() and, after that, call "AddTopTab".
But, now I'm calling Items.Clear() too, so we can remove this...

**Comment by felipebaltazar** [comment]
I`ve made the changes proposed by @PureWeen and updated this branch

**Comment by PureWeen** [comment]
@felipebaltazar can you rebase this to 4.4? 

Then I'll give the UI tests one more run through and merge.

Thank you!!

**Comment by felipebaltazar** [comment]
> @felipebaltazar can you rebase this to 4.4?
> 
> Then I'll give the UI tests one more run through and merge.
> 
> Thank you!!

@PureWeen I've rebased this branch to 4.4.0, but now we have 41 additional commits, seems like changes made on master.. so ca I only revert theses commits?

**Comment by PureWeen** [comment]
@felipebaltazar you'll need to rebase and force push. If you're having issues I can give it a go

git rebase --onto 4.4.0 master issue-6878

Then fix merge conflicts

Then do a  force push to your branch

**Comment by felipebaltazar** [comment]
> @felipebaltazar you'll need to rebase and force push. If you're having issues I can give it a go
> 
> git rebase --onto 4.4.0 master issue-6878
> 
> Then fix merge conflicts
> 
> Then do a force push to your branch

@PureWeen 
humm.. I already have made this and force pushed to this branch:

image

Btw, I've tried to do this again but seems like we have now 85 additional commits 🤔

**Comment by PureWeen** [comment]
@felipebaltazar looking through what's here it's a bit tricky to assess what went awry. For example (unless I'm missing something) I don't see the fix even being applied to *void ShellBottomNavViewAppearanceTracker.SetBackgroundColor* in these changes?


What might work best 
- create a new branch off of 4.4.0
- copy the necessary changes there
- create a new PR

Sorry about leading you astray on this one.

**Comment by felipebaltazar** [comment]
@PureWeen 
I've made a rollback to my first commit, at this PR, with `hard reset` and then I've made the changes from code review.

Now seems like ok to me...

Maybe we need to check these PR labels
**ps thanks for your patience

**Comment by PureWeen** [comment]
Yay! thank you

I'll kick the ui tests off and once they've passed I'll merge

**Comment by minahenin** [comment]
Hi, This is still broken in IOS

even when the Tab has multiple ShellItems when I try to Remove the first item "Items.RemoveAt(0);" it still crashes.

**Comment by francis2** [comment]
Hi guys,

I am using Forms Xamarin.4.5.0.282-pre4

I don't get crashes, but I am unable to remove the first item. 

These calls do nothing. The first Item always remains in the Items collection.
Items.RemoveAt(0)
Items.Remove(Tab.CurrentItem)
Items.Clear()

In my use case, the top menus are created based on the user data. I must create a default placeholder in xaml, but i need to delete it (or make it invisible, but that is not available yet).

    <TabBar>
        <Tab
            x:Name="mainTab"
            Title="SomeText"
            Route="someroute">
            <ShellContent
                Title="Empty"
                Route="blank">
                <ShellContent.ContentTemplate>
                    <DataTemplate>
                        <MyPage>
                            <x:Arguments>
                                <x:String></x:String>
                            </x:Arguments>
                        </MyPage>
                    </DataTemplate>
                </ShellContent.ContentTemplate>
            </ShellContent>            
        </Tab>
    </TabBar>

**Comment by PureWeen** [comment]
@francis2 do you have a repro you can attach?

**Comment by francis2** [comment]
Sure, here you go.

[PickerTestApp.zip](https://github.com/xamarin/Xamarin.Forms/files/4215264/PickerTestApp.zip)

**Comment by PureWeen** [comment]
Thank you @francis2 

https://github.com/xamarin/Xamarin.Forms/pull/9615

**Comment by francis2** [comment]
Hi guys,

Tab.Items.Clear() crashes on 4.5.0.356.

_System.ArgumentOutOfRangeException: Index was out of range. Must be non-negative and less than the size of the collection.
Parameter name: index_
  
Attached is a repro solution. Crash occurs in AppShell.xaml.cs line 28 on Tab.Items.Clear()

Thanks

[PickerTestApp.zip](https://github.com/xamarin/Xamarin.Forms/files/4277761/PickerTestApp.zip)

**Comment by PureWeen** [comment]
@francis2 

Created this issue for tracking
https://github.com/xamarin/Xamarin.Forms/issues/9801

I'll test and let you know over there what I find
